### ISSUE TITLE
Datatables, WET 4, and ajax data

### ISSUE THREAD

**Issue by NoisyMonk** [wet-boew/wet-boew#7838]
Hi all,

I've seen a few other posts about loading data via Ajax with JSON data, but I'm currently struggling to actually implement it.

One of the latest posts references using ["wb-init-loaded"](https://github.com/wet-boew/wet-boew/issues/2876), but that isn't reflected in the [current documentation](https://wet-boew.github.io/v4.0-ci/docs/ref/tables/tables-en.html).

The issue is that I'm not sure exactly when to try to load the ajax data. Either it's too early (before `$(document).on("wb-ready.wb-tables", ".wb-tables", function (event) { ...`) and I get "TypeError: $(...).DataTable is undefined" or I call it too late (after the above) and get "Cannot reinitialise DataTable..."

So my question is; when exactly do I call the following?

```
var table = $('#example').DataTable();
table.ajax.url('http://localhost:50538/Scripts/data.txt').load();
```

Thanks for your time.

**Comment by duboisp** [comment]
What about using this configuration instead of custom javascript:

you will need to replace the value of ```ajaxSource``` by your JSON file and adapt the columns mapping.

```html
<table class="wb-tables table table-striped table-hover" id="dataset-filter" aria-live="polite" data-wb-tables='{
            "bDeferRender": true,
            "ajaxSource": "ajax/datasource.json",
            "order": [5, "desc"],
             "columns": [
                { "data": "TITLE", "className": "nws-tbl-ttl h4" },
                { "data": "PUBDATE", "className": "nws-tbl-date" },
                { "data": "DEPT", "className": "nws-tbl-dept" },
                { "data": "TYPE", "className": "nws-tbl-type" },
                { "data": "TEASER",  "className": "nws-tbl-desc" },
                { "data": "LOCATION",  "visible": false },
                { "data": "AUDIENCE",  "visible": false },
                { "data": "SUBJECT", "visible": false },
                { "data": "MINISTER", "visible": false }
              ]}'>
 <thead>
  <tr>
   <th>Title</th>
   <th>Publication date</th>
   <th>Department</th>
   <th>News Type</th>
   <th>Summary</th>
   <th>Location</th>
   <th>For</th>
   <th>Subject</th>
   <th>Minister</th>
  </tr>
 </thead>
</table>
```

as the table filtering example, 3rd example in [data table example](http://wet-boew.github.io/wet-boew/demos/tables/tables-en.html)

**Comment by kodecount** [comment]
Also on another note if having issues their is a few quirks when loading a file directly vs ajax method I have experienced. One is hopefully being fix, but you can't have integer value in your JSON (integers need to be treated as strings). Also JSON must be formatted like (and start with the key `"data"`:

```
{
  "data": [ { "COLUMN": "DATA" } ]
}
```

**Comment by NoisyMonk** [comment]
@duboisp I was using an oversimplified example, that I was hoping I could extrapolate for my own needs afterwards, so I'll give a bit more detail. 

In my MVC 5 app I'm actually looking to load data from an (simplified) Action Method:

```
[HttpGet]
public ActionResult GetData()
{
    var viewModel = new ViewModel();
    viewModel = _queryService.GetData(100);

    var jsonData = new
    {
        data = viewModel
    };
    return Json(jsonData, JsonRequestBehavior.AllowGet);
}
```

So at some point (before, during, after?) the initialization of the grid, I need to load the data with an aJax call to that method.

Here's the table HTML ([taken from the DataTables site](https://datatables.net/examples/ajax/objects.html)):
```
<table id="example" class="display wb-tables table table-striped table-hover wb-tables-filter" cellspacing="0" width="100%">
    <thead>
        <tr>
            <th>Name</th>
            <th>Position</th>
            <th>Office</th>
            <th>Extn.</th>
            <th>Start date</th>
            <th>Salary</th>
        </tr>
    </thead>
    <tfoot>
        <tr>
            <th>Name</th>
            <th>Position</th>
            <th>Office</th>
            <th>Extn.</th>
            <th>Start date</th>
            <th>Salary</th>
        </tr>
    </tfoot>
</table>
```

And here's the closest I could find for the aJax call ([taken from the DataTables site](https://datatables.net/examples/ajax/objects.html)):
```
$("#example").DataTable({
    "ajax": "http://localhost:50538/GetData",
    "columns": [
        { "data": "name" },
        { "data": "position" },
        { "data": "office" },
        { "data": "extn" },
        { "data": "start_date" },
        { "data": "salary" }
    ]
});
```

@kodecount Yes, I've seen this limitation as well and I'm currently only using strings, but thanks for the reminder!

**Comment by duboisp** [comment]
@NoisyMonk can you post the schema of your JSON file or 2-3 lines of data returned by your function GetData()?

_FYI aside note - As per the HTML spec ```tfoot``` element is for column summary, not to replicate the table headers or not to contain table notes._ ;-)

**Comment by NoisyMonk** [comment]
@duboisp schema below - this is the sample test data from [DataTables](https://datatables.net/examples/ajax/objects.html).

Re: Footer: Noted! Also copy/part from [DataTables](https://datatables.net/examples/ajax/objects.html) though. :)

```
{
  "data": [
    {
      "name": "Tiger Nixon",
      "position": "System Architect",
      "salary": "$320,800",
      "start_date": "2011/04/25",
      "office": "Edinburgh",
      "extn": "5421"
    },
    {
      "name": "Garrett Winters",
      "position": "Accountant",
      "salary": "$170,750",
      "start_date": "2011/07/25",
      "office": "Tokyo",
      "extn": "8422"
    },
    ...
```

**Comment by kodecount** [comment]
Just my two cent. But I have tried to use `$("#example").DataTable({})` before wrapped in a `wb-ready` call and it bypassed WET data-table enhancements.

**Comment by duboisp** [comment]
This work for me and I used no javascript.

myPage.html
```html
<table id="example" class="display wb-tables table table-striped table-hover wb-tables-filter"
	 aria-live="polite" data-wb-tables='{
            "bDeferRender": true,
            "ajaxSource": "test.json",
             "columns": [
                 { "data": "name" },
                 { "data": "position" },
                 { "data": "office" },
                 { "data": "extn" },
                 { "data": "start_date" },
                 { "data": "salary" }
              ]}'>
    <thead>
        <tr>
            <th>Name</th>
            <th>Position</th>
            <th>Office</th>
            <th>Extn.</th>
            <th>Start date</th>
            <th>Salary</th>
        </tr>
    </thead>
</table>
```

test.json
```json
{
  "data": [
    {
      "name": "Tiger Nixon",
      "position": "System Architect",
      "salary": "$320,800",
      "start_date": "2011/04/25",
      "office": "Edinburgh",
      "extn": "5421"
    },
    {
      "name": "Garrett Winters",
      "position": "Accountant",
      "salary": "$170,750",
      "start_date": "2011/07/25",
      "office": "Tokyo",
      "extn": "8422"
	}
 ]
}
```

And that output this:
![issue-7838](https://cloud.githubusercontent.com/assets/1676030/21994831/afc3b4d8-dbef-11e6-8bf9-3e98b6094451.PNG)

Is the JSON return by the function ```public ActionResult GetData()``` are a valid JSON file?

**Comment by kodecount** [comment]
@duboisp Can `"ajaxSource": "test.json"` take a URL?

**Comment by duboisp** [comment]
@kodecount yes, I assume it can take any URL that is allowed by the cross-origin policy.

**Comment by duboisp** [comment]
Otherwise, I developed an alternative solution that use the [HTML5 template element and data-json](http://universallabs.org/wet/labs/data-json/v1.1-a2/templating.php). I expect that to be in the next release of WET where my PR should be submitted within a week or two.

**Comment by NoisyMonk** [comment]
@duboisp 
> Is the JSON return by the function public ActionResult GetData() are a valid JSON file?

Yep, the JSON returned is identical to the schema I posted earlier.

> This work for me and I used no javascript.

Yes, I can also get this to work but what I'm really after is a JS implementation of this so that I can dynamically change the data in the table. For example, what if I wanted to do a search, so I do a POST on some criteria and I want to return a new data set? I know DataTables can do this, but as soon as I try and use the WET 4 framework, I can't figure out how to dynamically load/change the table's data.

Surely someone has tried to do this with WET before?

**Comment by kodecount** [comment]
I have tried to dynamically change the table ran into issue of overriding WET DataTables enhancements. The simplest solution I found if you need a customized search (can't rely on the DataTables search) is to reload that page and rebuild with the new JSON string (using a POST or GET call).

**Comment by duboisp** [comment]
@NoisyMonk you can connect to DataTable by using something like:

```javascript
$datatable = $( "#" + $form.data( "bind-to" ) ).dataTable( { "retrieve": true } ).api();
```
([line 185 in data table plugin](https://github.com/wet-boew/wet-boew/blob/master/src/plugins/tables/tables.js#L185))

Then you should be able to do whatever is possible supported by the DataTable.

**Comment by NoisyMonk** [comment]
Unfortunately I couldn't spend any more time working on this so I abandoned that path.

Instead I just...

- destroy the table (`table.destroy()`)
- empty() it's div (`$('#grid').empty();`)
- do my ajax call to get the new data
- return my partial view grid (`$('#grid').html(result)`)
- reinitialize the table (`$(".wb-tables").trigger("wb-init.wb-tables");`)

Although it's not the answer I was looking for, it works. I'd consider this closed - thanks for everyone's help!

**Comment by kodecount** [comment]
You can also look at issue https://github.com/wet-boew/wet-boew/issues/7824 for other solutions.

**Comment by LaurentGoderre** [comment]
There is a section of the documentation of DataTables called server procesing which seem to do what you are looking for:

https://datatables.net/examples/data_sources/server_side.html
